# Moq1203: Method setup should specify a return value

| Item     | Value   |
| -------- | ------- |
| Enabled  | True    |
| Severity | Warning |
| CodeFix  | False   |
---

Method setups that have a return type should specify what value to return using `Returns()`, `ReturnsAsync()`, `Throws()`, or `ThrowsAsync()`.

When setting up methods that return values, Moq requires explicit specification of what should be returned or thrown. Without this specification, the setup is incomplete and may not behave as expected during testing. Intermediate calls like `Callback()` between `Setup()` and the return value specification do not affect the diagnostic.

## Examples of patterns that are flagged by this analyzer

```csharp
interface IFoo
{
    bool DoSomething(string value);
    int GetValue();
    Task<int> BarAsync();
}

var mock = new Mock<IFoo>();

// Moq1203: Method setup should specify a return value
mock.Setup(x => x.DoSomething("test"));

// Moq1203: Method setup should specify a return value
mock.Setup(x => x.GetValue());

// Moq1203: Async methods also need a return value specification
mock.Setup(x => x.BarAsync());

// Moq1203: Callback alone does not satisfy the requirement
mock.Setup(x => x.GetValue()).Callback(() => { });
```

## Solution

Specify what the method should return using `Returns()` or `ReturnsAsync()`, or what exception should be thrown using `Throws()` or `ThrowsAsync()`. These methods can follow `Setup()` at any position in the fluent chain, including after `Callback()`.

```csharp
interface IFoo
{
    bool DoSomething(string value);
    int GetValue();
    Task<int> BarAsync();
}

var mock = new Mock<IFoo>();

// Specify return value
mock.Setup(x => x.DoSomething("test")).Returns(true);
mock.Setup(x => x.GetValue()).Returns(42);

// Async methods can use ReturnsAsync
mock.Setup(x => x.BarAsync()).ReturnsAsync(1);

// Callback before Returns is valid
mock.Setup(x => x.DoSomething("test")).Callback(() => { }).Returns(true);
mock.Setup(x => x.BarAsync()).Callback(() => { }).ReturnsAsync(1);

// Specify exception to throw
mock.Setup(x => x.DoSomething("test")).Throws<InvalidOperationException>();
mock.Setup(x => x.BarAsync()).ThrowsAsync(new InvalidOperationException());
```

## When this rule does not apply

This analyzer only applies to methods that have a return type. The following scenarios are not flagged:

- **Void methods**: Methods that return `void` do not need return value specification
- **Property setups**: Properties use different setup methods (`SetupGet`, `SetupSet`, `SetupProperty`)

```csharp
interface IFoo
{
    void DoVoidMethod();
    void ProcessData(string data);
    string Name { get; set; }
}

var mock = new Mock<IFoo>();

// These are valid and not flagged:
mock.Setup(x => x.DoVoidMethod());           // Void method
mock.Setup(x => x.ProcessData("test"));      // Void method
mock.Setup(x => x.Name);                     // Property (handled differently)
mock.SetupGet(x => x.Name);                  // Proper property setup
```

## Suppress a warning

If you just want to suppress a single violation, add preprocessor directives to
your source file to disable and then re-enable the rule.

```csharp
#pragma warning disable Moq1203
mock.Setup(x => x.GetValue()); // Moq1203: Method setup should specify a return value
#pragma warning restore Moq1203
```

To disable the rule for a file, folder, or project, set its severity to `none`
in the
[configuration file](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/configuration-files).

```ini
[*.{cs,vb}]
dotnet_diagnostic.Moq1203.severity = none
```

For more information, see
[How to suppress code analysis warnings](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/suppress-warnings).