# Moq1208: Returns() delegate type mismatch on async method setup

| Item     | Value |
| -------- | ----- |
| Enabled  | True  |
| Severity | Warning |
| CodeFix  | True |
---

When `.Returns()` is called with a delegate that returns a non-Task type on an async method setup, Moq throws a runtime `MockException`. The delegate return type must match the method return type. For async methods returning `Task<T>` or `ValueTask<T>`, the delegate must return `Task<T>` or `ValueTask<T>`, not `T` directly.

```
MockException: Invalid callback. Setup on method with return type 'Task' cannot invoke callback with return type 'int'.
```

## Examples of patterns that are flagged by this analyzer

```csharp
public interface IService
{
    Task<int> GetValueAsync();
    Task<string> GetNameAsync();
    ValueTask<int> GetValueTaskAsync();
}

var mock = new Mock<IService>();

// These patterns are flagged:
mock.Setup(x => x.GetValueAsync()).Returns(() => 42); // Moq1208: Returns() delegate type mismatch on async method setup
mock.Setup(x => x.GetNameAsync()).Returns(() => "hello"); // Moq1208: Returns() delegate type mismatch on async method setup
mock.Setup(x => x.GetValueTaskAsync()).Returns(() => 42); // Moq1208: Returns() delegate type mismatch on async method setup
```

## Solution

```csharp
public interface IService
{
    Task<int> GetValueAsync();
    Task<string> GetNameAsync();
    ValueTask<int> GetValueTaskAsync();
}

var mock = new Mock<IService>();

// Use ReturnsAsync instead (recommended):
mock.Setup(x => x.GetValueAsync()).ReturnsAsync(42);
mock.Setup(x => x.GetValueAsync()).ReturnsAsync(() => 42);

// Or return Task explicitly:
mock.Setup(x => x.GetValueAsync()).Returns(() => Task.FromResult(42));
mock.Setup(x => x.GetNameAsync()).Returns(() => Task.FromResult("hello"));
mock.Setup(x => x.GetValueTaskAsync()).Returns(() => new ValueTask<int>(42));
```

## Suppress a warning

If you just want to suppress a single violation, add preprocessor directives to
your source file to disable and then re-enable the rule.

```csharp
#pragma warning disable Moq1208
mock.Setup(x => x.GetValueAsync()).Returns(() => 42); // Moq1208: Returns() delegate type mismatch on async method setup
#pragma warning restore Moq1208
```

To disable the rule for a file, folder, or project, set its severity to `none`
in the
[configuration file](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/configuration-files).

```ini
[*.{cs,vb}]
dotnet_diagnostic.Moq1208.severity = none
```

For more information, see
[How to suppress code analysis warnings](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/suppress-warnings).
