# Moq1204: Raises event arguments should match the event delegate signature

## Cause

Arguments passed to `Raises()` do not match the event delegate signature.

## Rule description

The `Raises` method allows you to simulate an event being fired when a mocked method is called. For example, you can configure a `Submit()` method to automatically trigger a `Completed` event.

To identify the event, `Raises` takes a lambda expression like `x => x.Completed += null`. The `+= null` part is the standard Moq syntax for targeting an event without actually providing a handler.

This rule ensures that the arguments you provide to `Raises` are compatible with the signature of the event's delegate. For instance, if an event is defined as `event Action<string>`, the `Raises` method must be called with a `string` argument.

**Why this is important:** If the arguments do not match the event's delegate signature, your test code will compile successfully. However, the test will fail at runtime when Moq attempts to invoke the event with mismatched arguments. This analyzer turns a subtle runtime bug into a clear compile-time error, making it easier to write correct tests.

## How to fix violations

Ensure that the arguments passed to `Raises()` match the types expected by the event delegate signature.

## When to suppress warnings

Do not suppress warnings from this rule. Incorrect event arguments will cause runtime exceptions.

## Example of a violation

```csharp
interface INotifier
{
    void Submit();
    event Action<string> Completed;
}

class Test
{
    void TestMethod()
    {
        var mock = new Mock<INotifier>();

        // Wrong: passing int instead of string
        mock.Setup(x => x.Submit()).Raises(x => x.Completed += null, 42);
    }
}
```

## Example of how to fix

```csharp
interface INotifier
{
    void Submit();
    event Action<string> Completed;
}

class Test
{
    void TestMethod()
    {
        var mock = new Mock<INotifier>();

        // Correct: passing string as expected by Action<string>
        mock.Setup(x => x.Submit()).Raises(x => x.Completed += null, "success");
    }
}
```

## Supported event delegate types

- `Action` (no parameters)
- `Action<T>`, `Action<T1, T2>`, etc. (generic Action delegates)
- `EventHandler<T>` (expects single argument of type T)
- Custom delegate types (analyzed via their Invoke method signature)

## Compatibility

This rule applies to all supported Moq versions as the `Raises` method has been available since early versions.
