# Moq1203: Event setup handler type should match event delegate type

## Cause

Handler type passed to `SetupAdd` or `SetupRemove` does not match the event delegate type.

## Rule description

When you set up an event on a mocked object using `SetupAdd` or `SetupRemove`, you are telling Moq how to behave when code subscribes to or unsubscribes from that event. A common way to do this is to specify the type of delegate that can be added or removed, often using `It.IsAny<T>()`.

This rule ensures that the delegate type you specify in your setup (e.g., `It.IsAny<Action<int>>`) is compatible with the actual delegate type of the event (e.g., `event Action<string>`).

**Why this is important:** If the types do not match, your test code will compile without errors. However, when the test runs and the event is accessed, Moq will throw a runtime exception because it cannot handle the incorrect delegate type. This analyzer prevents that runtime error by flagging the mismatch during development, saving you from debugging unexpected test failures.

## How to fix violations

Ensure that the handler type passed to `SetupAdd` or `SetupRemove` matches the event's delegate type.

## When to suppress warnings

Do not suppress warnings from this rule. Incorrect handler types will cause runtime exceptions.

## Example of a violation

```csharp
interface INotifier
{
    event Action<string> StringChanged;
}

class Test
{
    void TestMethod()
    {
        var mock = new Mock<INotifier>();
        
        // Wrong: passing Action<int> instead of Action<string>
        mock.SetupAdd(x => x.StringChanged += It.IsAny<Action<int>>());
    }
}
```

## Example of how to fix

```csharp
interface INotifier
{
    event Action<string> StringChanged;
}

class Test
{
    void TestMethod()
    {
        var mock = new Mock<INotifier>();
        
        // Correct: passing Action<string> as expected by the event
        mock.SetupAdd(x => x.StringChanged += It.IsAny<Action<string>>());
    }
}
```

## Supported event delegate types

- `Action` (no parameters)
- `Action<T>`, `Action<T1, T2>`, etc. (generic Action delegates)
- `EventHandler<T>` (standard .NET event handler pattern)
- Custom delegate types (analyzed via their signature)

## Compatibility

This rule only applies to Moq 4.18.4+ as `SetupAdd` and `SetupRemove` are not available in earlier versions.