# Moq1204: Raises event arguments should match the event delegate signature

## Cause

Arguments passed to `Raises()` do not match the event delegate signature.

## Rule description

When using `Raises()` to trigger events as part of method setup, the arguments provided must match the parameter types expected by the event delegate.

## How to fix violations

Ensure that the arguments passed to `Raises()` match the types expected by the event delegate signature.

## When to suppress warnings

Do not suppress warnings from this rule. Incorrect event arguments will cause runtime exceptions.

## Example of a violation

```csharp
interface INotifier
{
    void Submit();
    event Action<string> Completed;
}

class Test
{
    void TestMethod()
    {
        var mock = new Mock<INotifier>();
        
        // Wrong: passing int instead of string
        mock.Setup(x => x.Submit()).Raises(x => x.Completed += null, 42);
    }
}
```

## Example of how to fix

```csharp
interface INotifier
{
    void Submit();
    event Action<string> Completed;
}

class Test
{
    void TestMethod()
    {
        var mock = new Mock<INotifier>();
        
        // Correct: passing string as expected by Action<string>
        mock.Setup(x => x.Submit()).Raises(x => x.Completed += null, "success");
    }
}
```

## Supported event delegate types

- `Action` (no parameters)
- `Action<T>`, `Action<T1, T2>`, etc. (generic Action delegates)
- `EventHandler<T>` (expects single argument of type T)
- Custom delegate types (analyzed via their Invoke method signature)

## Compatibility

This rule applies to all supported Moq versions as the `Raises` method has been available since early versions.