# Moq1202: Raise event arguments should match the event delegate signature

## Cause

Arguments passed to `Mock.Raise()` do not match the event delegate signature.

## Rule description

When using `Mock.Raise()` to trigger events on mocked objects, the arguments provided must match the parameter types expected by the event delegate.

## How to fix violations

Ensure that the arguments passed to `Mock.Raise()` match the types expected by the event delegate signature.

## When to suppress warnings

Do not suppress warnings from this rule. Incorrect event arguments will cause runtime exceptions.

## Example of a violation

```csharp
interface IOptionsProvider<TOptions>
{
    event Action<TOptions> OptionsChanged;
}

class Test
{
    void TestMethod()
    {
        var mock = new Mock<IOptionsProvider<string>>();
        
        // Wrong: passing int instead of string
        mock.Raise(p => p.OptionsChanged += null, 42);
    }
}
```

## Example of how to fix

```csharp
interface IOptionsProvider<TOptions>
{
    event Action<TOptions> OptionsChanged;
}

class Test
{
    void TestMethod()
    {
        var mock = new Mock<IOptionsProvider<string>>();
        
        // Correct: passing string as expected by Action<string>
        mock.Raise(p => p.OptionsChanged += null, "correct value");
    }
}
```

## Supported event delegate types

- `Action` (no parameters)
- `Action<T>`, `Action<T1, T2>`, etc. (generic Action delegates)
- `EventHandler<T>` (expects single argument of type T)
- Custom delegate types (analyzed via their Invoke method signature)