# Moq1100: Callback signature must match the signature of the mocked method

| Item     | Value   |
| -------- | ------- |
| Enabled  | True    |
| Severity | Warning |
| CodeFix  | True    |

---

The signature of the `.Callback()` method must match the signature of the `.Setup()` method. To fix:

- Ensure the parameters to `.Callback()` match the signature created by `.Setup()`. A code fix is available to automatically
  match

This rule supports both lambda expressions and delegate constructor syntax for callbacks.

## Examples of patterns that are flagged by this analyzer

```csharp
interface IMyService
{
    int Do(int i, string s, DateTime dt);
}

var mock = new Mock<IMyService>()
    .Setup(x => x.Do(It.IsAny<int>(), It.IsAny<string>(), It.IsAny<DateTime>()))
    .Callback((string s1, int i1) => { }); // Moq1100: Callback signature must match the signature of the mocked method
```

### Delegate constructor callbacks

```csharp
public delegate void DoCallback(int i, string s, DateTime dt);

interface IMyService
{
    int Do(int i, string s, DateTime dt);
}

var mock = new Mock<IMyService>()
    .Setup(x => x.Do(It.IsAny<int>(), It.IsAny<string>(), It.IsAny<DateTime>()))
    .Callback(new DoCallback((string s1, int i1, DateTime dt1) => { })); // Moq1100: Callback signature must match the signature of the mocked method
```

## Solution

```csharp
interface IMyService
{
    int Do(int i, string s, DateTime dt);
}

var mock = new Mock<IMyService>()
    .Setup(x => x.Do(It.IsAny<int>(), It.IsAny<string>(), It.IsAny<DateTime>()))
    .Callback((int i, string s, DateTime dt) => { });
```

### With delegate constructor

```csharp
public delegate void DoCallback(int i, string s, DateTime dt);

interface IMyService
{
    int Do(int i, string s, DateTime dt);
}

var mock = new Mock<IMyService>()
    .Setup(x => x.Do(It.IsAny<int>(), It.IsAny<string>(), It.IsAny<DateTime>()))
    .Callback(new DoCallback((int i, string s, DateTime dt) => { }));
```

## Advanced Patterns Supported

This analyzer supports comprehensive validation of advanced callback patterns:

### Multiple Callback Timing

```csharp
mock.Setup(x => x.DoWork("test"))
    .Callback(() => Console.WriteLine("Before"))
    .Returns(42)
    .Callback(() => Console.WriteLine("After"));
```

### Ref/Out Parameter Callbacks

```csharp
delegate void ProcessDataCallback(ref string data);
mock.Setup(x => x.ProcessData(ref It.Ref<string>.IsAny))
    .Callback(new ProcessDataCallback((ref string data) => data = "processed"));
```

### Complex Multi-Parameter Scenarios

```csharp
mock.Setup(x => x.ProcessMultiple(It.IsAny<int>(), It.IsAny<string>(), It.IsAny<DateTime>()))
    .Callback((int id, string name, DateTime timestamp) => Console.WriteLine($"Processing {id}"));
```

### Out Parameter Delegates

```csharp
delegate bool TryProcessCallback(out int result);
mock.Setup(x => x.TryProcess(out It.Ref<int>.IsAny))
    .Callback(new TryProcessCallback((out int result) => { result = 42; }))
    .Returns(true);
```

## Known Limitations

### Generic Callback Syntax

This analyzer does **not** currently validate type parameter mismatches when using the explicit generic `.Callback<T>()` syntax. For example:

```csharp
// This will NOT trigger a diagnostic (current limitation)
mock.Setup(x => x.DoWork("test")) // DoWork takes a string parameter
    .Callback<int>(wrongTypeParam => { }); // Using int instead of string - no warning!
```

**Best Practice:** Use lambda parameter inference instead of explicit generic syntax to get full validation:

```csharp
// ✅ Recommended: Let the compiler infer parameter types
mock.Setup(x => x.DoWork("test"))
    .Callback(param => { }); // Type is inferred correctly as string

// ✅ Also recommended: Explicitly type the lambda parameter
mock.Setup(x => x.DoWork("test"))
    .Callback((string param) => { }); // Type validation works correctly
```

**Rationale:** The explicit generic `.Callback<T>()` syntax is rarely used in practice. Analysis of open-source Moq usage found zero instances of this pattern. The recommended lambda parameter inference approach provides full type safety and validation coverage.

## Suppress a warning

If you just want to suppress a single violation, add preprocessor directives to
your source file to disable and then re-enable the rule.

```csharp
#pragma warning disable Moq1100
var mock = new Mock<IMyService>()
    .Setup(x => x.Do(It.IsAny<int>(), It.IsAny<string>(), It.IsAny<DateTime>()))
    .Callback((string s1, int i1) => { }); // Moq1100: Callback signature must match the signature of the mocked method
#pragma warning restore Moq1100
```

To disable the rule for a file, folder, or project, set its severity to `none`
in the
[configuration file](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/configuration-files).

```ini
[*.{cs,vb}]
dotnet_diagnostic.Moq1100.severity = none
```

For more information, see
[How to suppress code analysis warnings](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/suppress-warnings).
